[RFC] Added context-pure mixin and tests

hai-cea opened the pull request:
This mixin checks for context changes in addition to prop and state changes in `shouldComponentUpdate` and is an attempt to address #1176.

The `ContextPure` mixin requires you to declare which context props you care about. To do this, you can add one or both of the following static functions:


#### getContextProps
If the component is dependent on certain context variables, declare them in a static function named getContextProps. This function will take in a context object parameter and return an object containing the context variables needed to render the component. This function is used in `shouldComponentUpdate` to check for context changes. For example:
```javascript
statics: {
  getContextProps(context) {
    return {
      fontFamily: context.muiTheme.contentFontFamily,
    }
  },
```


#### getChildrenClasses
If the component also renders other material-ui components, we'll also need to declare these child components in a static function named getChildrenClasses. This function should return an array of react components that the component renders.  This will allow us to trigger an update in `shouldComponentUpdate` when children context variables change. For example:
```javascript
statics: {
  getChildrenClasses() {
    return [
      EnhancedButton,
      Paper,
    ];
  },
}
```

cgestes commented:
Good idea. 
Looks nice :)

oliviertassinari commented:
Yeah, can't wait to see this in action.

The check of context can be pretty expensive when the component has a lot of subcomponents, for instance the DatePicker.
What do you think about not calling `contextPropsEqual` in `shouldComponentUpdate` if `nextContext` contains a specfic property equal to `true`. Let's say `contextPureDisabled`.

hai-cea commented:
@oliviertassinari Yeh, I think that's a good idea. I'll update the PR.

hai-cea commented:
@oliviertassinari I added a check for `muiTheme.static`.

If you set `muiTheme.static` to `true`, it will bypass the `contextPropsEqual` check.

oliviertassinari commented:
Awesome

hai-cea commented:
@oliviertassinari Good Idea - I updated the ThemeManager...which I think needs an overhaul (but we'll save that for another time :)) Also updated tests to use the ThemeManager.

oliviertassinari commented:
@hai-cea Perfect :smiley: . We will have to add in the release note that static `true` introduce a breaking changes.

shaurya947 commented:
Great idea, but I strongly feel that the method name `getContextProps()` is a little misleading since context and props are both entities in a React component, and we're using one as an adjective for the other. I think we should rename it to something like `getKeysFromContext()` or `getRelevantContextKeys()` or something along those lines which conveys the meaning better. I know @hai-cea is not too intent on the idea due to potential breaking changes. But I think it's still worth doing considering that very little code actually uses this functionality today, and changing it could make the meaning much clearer.

What do you think @oliviertassinari @cgestes ?

cgestes commented:
I do think the change make sense. :)

On Wed, Sep 16, 2015 at 10:17 PM Shaurya Arora <notifications@github.com>
wrote:

> Great idea, but I strongly feel that the method name getContextProps() is
> a little misleading since context and props are both entities in a React
> component, and we're using one as an adjective for the other. I think we
> should rename it to something like getKeysFromContext() or
> getRelevantContextKeys() or something along those lines which conveys the
> meaning better. I know @hai-cea <https://github.com/hai-cea> is not too
> intent on the idea due to potential breaking changes. But I think it's
> still worth doing considering that very little code actually uses this
> functionality today, and changing it could make the meaning much clearer.
>
> What do you think @oliviertassinari <https://github.com/oliviertassinari>
> @cgestes <https://github.com/cgestes> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/callemall/material-ui/pull/1397#issuecomment-140873939>
> .
>

oliviertassinari commented:
Yeah, I agree
